#  Recomendador de pelis

Luis Carlos Marrufo Padilla

Angel Esparza Enriquez

Gustavo Andres Aguilar Torreblanca

Jose Luis Santos Montaño

![](http://labs.criteo.com/wp-content/uploads/2017/08/CustomersWhoBought3.jpg)

En este cuaderno, se busca  implementar algunos algoritmos de recomendación (basados en contenido, en popularidad y en filtrado colaborativo) y se va a construir un conjunto (ensemble) de estos modelos para crear nuestro sistema de recomendación final. Contamos con dos conjuntos de datos de MovieLens.

El conjunto de datos completo: Consiste en 26,000,000 de calificaciones y 750,000 aplicaciones de etiquetas aplicadas a 45,000 películas por 270,000 usuarios. Incluye datos del genoma de etiquetas con 12 millones de puntuaciones de relevancia en 1,100 etiquetas.

El conjunto de datos pequeño: Comprende 100,000 calificaciones y 1,300 aplicaciones de etiquetas aplicadas a 9,000 películas por 700 usuarios.

Construye un **Recomendador Simple** utilizando películas del conjunto de datos completo, mientras que todos los sistemas de recomendación personalizados harán uso del conjunto de datos pequeño (debido a que mi capacidad de cómputo es muy limitada). Como primer paso, construiré mi sistema de recomendación simple.

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

import warnings
warnings.simplefilter('ignore')




### Análisis de datos y manipulación
import pandas as pd  # Estructuras de datos flexibles y herramientas de análisis (DataFrames, Series)

import numpy as np   # Computación numérica con arrays multidimensionales y funciones matemáticas

### Visualización de datos
import matplotlib.pyplot as plt  # Biblioteca fundamental para crear visualizaciones estáticas, animadas e interactivas

import seaborn as sns            # Biblioteca de visualización basada en matplotlib con estilos atractivos y gráficos estadísticos

### Análisis estadístico
from scipy import stats  # Funciones estadísticas (distribuciones, tests estadísticos, medidas de correlación)

### Evaluación segura de strings
from ast import literal_eval  # Convierte strings que contienen estructuras de Python en las estructuras mismas

### Procesamiento de texto y machine learning
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer  # Conversión de texto a vectores numéricos

from sklearn.metrics.pairwise import linear_kernel, cosine_similarity  # Cálculo de similitudes entre vectores

### Procesamiento de lenguaje natural (NLP)
from nltk.stem.snowball import SnowballStemmer  # Algoritmo de stemming para reducir palabras a su raíz

from nltk.stem.wordnet import WordNetLemmatizer # Lematización para reducir palabras a su forma canónica

from nltk.corpus import wordnet                 # Base de datos léxica del inglés para lematización

### Sistema de recomendación
from surprise import Reader, Dataset, SVD       # Biblioteca para sistemas de recomendación (Reader: formato de datos, Dataset: conjunto de datos, SVD: Descomposición en Valores Singulares)

from surprise.model_selection import cross_validate  # Validación cruzada para modelos de recomendación


## Recomendador Simple 

El Recomendador Simple ofrece recomendaciones generales a todos los usuarios basadas en la popularidad de las películas y (a veces) en su género. La idea básica detrás de este recomendador es que las películas más populares y con mejores críticas tienen una mayor probabilidad de ser del agrado del público promedio. Este modelo no proporciona recomendaciones personalizadas basadas en el usuario (como nuestro reto).

La implementación de este modelo es extremadamente sencilla. Todo lo que debemos hacer es ordenar las películas según sus calificaciones y popularidad, y mostrar las películas principales de la lista. Como paso adicional, podemos incluir un argumento de género para obtener las mejores películas de un género en particular.

Cargar los datos

In [2]:
md = pd.read_csv('data/movies_metadata.csv')
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

Vamos a trabajar por género

In [3]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

Usaremos las calificaciones de TMDb para crear nuestro Ranking de las Mejores Películas (Top Movies Chart). La fórmula de calificación ponderada de IMDB para construir este ranking. Matemáticamente, se representa de la siguiente manera:

Calificación Ponderada (WR) = $(\frac{v}{v + m} \cdot R) + (\frac{m}{v + m} \cdot C)$

donde,

v es el número de votos que tiene la película

m es el número mínimo de votos requeridos para ser incluida en el ranking

R es la calificación promedio de la película

C es la calificación promedio de todas las películas del conjunto de datos

El siguiente paso es determinar un valor apropiado para m, el número mínimo de votos necesarios para aparecer en el ranking. Usaremos el percentil 95 como punto de corte. En otras palabras, para que una película aparezca en el ranking, debe tener más votos que al menos el 95% de las películas en la lista.

El ranking general de las 250 mejores películas se define una función para crear rankings por género específico. 

In [4]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [5]:
m = vote_counts.quantile(0.95)
m

434.0

Extracción del año:

In [6]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

Vamos a trabajar con el siguiente data frame:

In [7]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

Por lo tanto, para calificar y ser considerada en el ranking, una película debe tener al menos ----- votos en TMDB. También observamos que la calificación promedio de una película en TMDB es de --- en una escala de 10. En total, ---- películas califican para estar en nuestro ranking.

In [8]:
def weighted_rating(x):
    # Función de Calificación ponderada
    v = x['vote_count']
    R = x['vote_average']
    
    return (v / (v + m) * R) + (m / (v + m) * C)

In [9]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [10]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

### Películas Top 

In [11]:
#Imprime las primeras 15 pelis del rank

qualified.head(15)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1225                                  Back to the Future  1985        6239   
834                                        The Godfather  1972        6024   
1154                             The Empire Strikes Back  1980        5998   
46                                                 Se7en  1995        5915   

       vote_average  popularity  \
15480             8   29.108149   
12481             8  123.167259   
22879             8   32.213481   
2843              8   63.869599   
4863              8   32.070725   
292               8  140.950236   
314               8   51.645403   
7000              8   29.324358   
351               8   48.307194   
5814              8   29.423537   
256               8   42.149697   
1225              8   25.778509   
834               8   41.109264   
1154              8   19.470959   
46                8    18.45743   

                                                  genres        wr  
15480  [Action, Thriller, Science Fiction, Mystery, A...  7.917588  
12481                   [Drama, Action, Crime, Thriller]  7.905871  
22879                [Adventure, Drama, Science Fiction]  7.897107  
2843                                             [Drama]  7.881753  
4863                        [Adventure, Fantasy, Action]  7.871787  
292                                    [Thriller, Crime]  7.868660  
314                                       [Drama, Crime]  7.864000  
7000                        [Adventure, Fantasy, Action]  7.861927  
351                             [Comedy, Drama, Romance]  7.860656  
5814                        [Adventure, Fantasy, Action]  7.851924  
256                 [Adventure, Action, Science Fiction]  7.834205  
1225        [Adventure, Comedy, Science Fiction, Family]  7.820813  
834                                       [Drama, Crime]  7.814847  
1154                [Adventure, Action, Science Fiction]  7.814099  
46                            [Crime, Mystery, Thriller]  7.811669

Vemos que tres películas de Christopher Nolan —Inception, The Dark Knight e Interstellar, aparecen en los primeros lugares de nuestro ranking. El gráfico también muestra una fuerte preferencia de los usuarios de TMDB hacia ciertos géneros y directores.

Ahora construiremos una función que genere rankings para géneros específicos. Para ello, relajaremos nuestras condiciones por defecto utilizando el percentil 85 en lugar del 95.

In [12]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)

In [13]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

Veamos el método mostrando el Top 15 de películas de romance (el género de romance casi no apareció en nuestro ranking general, a pesar de ser uno de los géneros cinematográficos más populares).

### Top 15 de películas de Romance 

In [14]:
build_chart('Romance').head(15)

title  year  vote_count  vote_average popularity  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9  34.457024   
351                   Forrest Gump  1994        8147             8  48.307194   
876                        Vertigo  1958        1162             8   18.20822   
40251                   Your Name.  2016        1030             8  34.461252   
883               Some Like It Hot  1959         835             8  11.845107   
1132               Cinema Paradiso  1988         834             8  14.177005   
19901                     Paperman  2012         734             8   7.198633   
37863                  Sing Street  2016         669             8  10.672862   
882                  The Apartment  1960         498             8  11.994281   
38718               The Handmaiden  2016         453             8  16.727405   
3189                   City Lights  1931         444             8  10.891524   
24886             The Way He Looks  2014         262             8   5.711274   
45437               In a Heartbeat  2017         146             8   20.82178   
1639                       Titanic  1997        7770             7   26.88907   
19731      Silver Linings Playbook  2012        4840             7  14.488111   

             wr  
10309  8.565285  
351    7.971357  
876    7.811667  
40251  7.789489  
883    7.745154  
1132   7.744878  
19901  7.713951  
37863  7.689483  
882    7.599317  
38718  7.566166  
3189   7.558867  
24886  7.331363  
45437  7.003959  
1639   6.981546  
19731  6.970581

La mejor película de romance según nuestras métricas es la producción de Bollywood Dilwale Dulhania Le Jayenge. Esta cinta protagonizada por Shahrukh Khan.

## Recomendador Basado en Contenido

El recomendador que construimos en la sección anterior presenta algunas limitaciones importantes. Por ejemplo, ofrece las mismas recomendaciones a todos los usuarios, sin tener en cuenta sus gustos personales. Si una persona que ama las películas románticas (y detesta la acción) revisará nuestro Top 15, probablemente no disfrutaría de la mayoría de las películas. Incluso si revisara nuestros rankings por género, aún no estaría recibiendo las mejores recomendaciones.

Por ejemplo, consideremos a una persona que ama *Dilwale Dulhania Le Jayenge*, *My Name is Khan* y *Kabhi Khushi Kabhi Gham*. Podemos inferir que esta persona disfruta de las películas protagonizadas por el actor **Shahrukh Khan** y dirigidas por **Karan Johar**. Aun si consultara el ranking de películas románticas, probablemente no encontraría estas entre las principales recomendaciones.

Para personalizar mejor nuestras recomendaciones, hay que construir un motor que calcule la **similitud entre películas** basándose en ciertos atributos, y que sugiera aquellas más parecidas a una película que le haya gustado al usuario. Dado que utilizaremos metadatos de las películas (o su contenido) para construir este motor, este enfoque se conoce como **Filtrado Basado en Contenido**.

Vamops a construir dos recomendadores basados en contenido, a partir de:
* Sinopsis y eslóganes de las películas  
* Reparto, equipo técnico, palabras clave y género  

Como se mencionó en la introducción, se utiliza un subconjunto de todas las películas disponibles debido a las limitaciones de capacidad de cómputo.


In [15]:
links_small = pd.read_csv('data/links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [16]:
md = md.drop([19730, 29503, 35587])

In [17]:
md['id'] = md['id'].astype('int')

In [18]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9099, 25)

 Tenemos **Completa** películas disponibles en nuestro conjunto de datos reducido de metadatos de películas,
 el cual es **Completa** veces más pequeño que nuestro conjunto de datos original de 45,000 películas.


### Recomendador Basado en la Descripción de la Película

Primero intentemos construir un recomendador utilizando las **descripciones** y **eslóganes** de las películas.  
No contamos con una métrica cuantitativa para evaluar el rendimiento de nuestro modelo, por lo que esta evaluación deberá realizarse de forma **cualitativa**.


In [19]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

Se analiza las palabras en la descripción, creando unigramas y bigramas

Películas con descripciones similares tendrán vectores TF-IDF similares

In [20]:
tf = TfidfVectorizer(analyzer='word', ngram_range=(1, 2), min_df=1, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])

In [21]:
tfidf_matrix.shape

(9099, 268124)

#### Similitud del Coseno

Vamos a utilizar la **Similitud del Coseno** para calcular una cantidad numérica que indique el grado de similitud entre dos películas.  
Matemáticamente, se define de la siguiente manera:

$cosine(x,y) = \frac{x. y^\intercal}{||x||.||y||} $

Dado que hemos utilizado el **Vectorizer TF-IDF**, calcular el **producto punto** nos dará directamente el valor de la similitud del coseno.  
Por lo tanto, utilizaremos la función **linear_kernel** de *sklearn* en lugar de *cosine_similarities*, ya que es más rápida.


In [22]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [23]:
cosine_sim[0]

array([1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
       0.        ])

Ahora tenemos una **matriz de similitud del coseno por pares** para todas las películas de nuestro conjunto de datos.  
El siguiente paso es escribir una función que devuelva las **30 películas más similares** basándose en el puntaje de similitud del coseno.


In [24]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [25]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

Ahora intentemos obtener las **principales recomendaciones** para algunas películas y ver qué tan buenas son las sugerencias.


In [26]:
get_recommendations('The Godfather').head(10) # completa para las primeras 10 

973      The Godfather: Part II
8387                 The Family
3509                       Made
4196         Johnny Dangerously
29               Shanghai Triad
5667                       Fury
2412             American Movie
1582    The Godfather: Part III
4221                    8 Women
2159              Summer of Sam
Name: title, dtype: object

In [27]:
get_recommendations('The Dark Knight').head(10) # completa para las primeras 10 

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
2696                                        JFK
8165    Batman: The Dark Knight Returns, Part 1
Name: title, dtype: object

Podemos observar que para **The Dark Knight**, nuestro sistema logra identificarla como una película de *Batman* y, en consecuencia, recomienda otras películas de *Batman* como sus principales sugerencias.  
Sin embargo, lamentablemente eso es todo lo que este sistema puede hacer por el momento.  
Esto no resulta muy útil para la mayoría de las personas, ya que no toma en cuenta características muy importantes como el **elenco**, **equipo técnico**, **director** y **género**, los cuales influyen en la calificación y popularidad de una película.  
Alguien a quien le gustó **The Dark Knight** probablemente la disfruta más por **Nolan**, y seguramente detestaría **Batman Forever** y otras películas de menor calidad dentro de la franquicia de Batman.

Por lo tanto, pueden estudiar los  metadatos,  que pueden ser más representativos que solo la **sinopsis** y el **eslogan**.  
En la siguiente subsección construiremos un recomendador más sofisticado que tenga en cuenta el **género**, las **palabras clave**, el **reparto** y el **equipo técnico**.


### Recomendador Basado en Metadatos

Para construir nuestro recomendador estándar basado en contenido y metadatos, necesitaremos combinar nuestro conjunto de datos actual con los conjuntos de datos de **equipo técnico (crew)** y **palabras clave (keywords)**.  
Prepararemos estos datos como **primer paso**.


In [28]:
credits = pd.read_csv('data/credits.csv')
keywords = pd.read_csv('data/keywords.csv')

In [29]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

In [30]:
md.shape

(45463, 25)

In [31]:
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [32]:
smd = md[md['id'].isin(links_small)]
smd.shape

(9219, 28)

Ahora tenemos el **reparto**, el **equipo técnico**, los **géneros** y los **créditos**, todos en un mismo *dataframe*.  
Vamos a procesar estos datos un poco más utilizando las siguientes intuiciones:

1. **Equipo técnico (Crew):**  
   Del equipo técnico, solo tomaremos al **director** como característica, ya que los demás miembros no influyen tanto en la *esencia* de la película.

2. **Reparto (Cast):**  
   La selección del reparto es un poco más complicada.  
   Los actores poco conocidos o los papeles menores no afectan demasiado la opinión del público sobre una película.  
   Por lo tanto, debemos seleccionar solo los **personajes principales** y sus respectivos actores.  
   De manera arbitraria, elegiremos a los **3 actores principales** que aparecen en la lista de créditos.


In [33]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

Extraemos a los Directores

In [34]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [35]:
smd['director'] = smd['crew'].apply(get_director)

Extraemos a los actores

In [36]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [37]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

El enfoque para construir este recomendador va a ser *experimental*.  
Se planea crear un **volcado de metadatos** (*metadata dump*) para cada película, que contenga los siguientes elementos:  
**géneros, director, actores principales y palabras clave.**  
Luego, se utiliza un **Count Vectorizer** para generar nuestra matriz de conteo, tal como hicimos en el recomendador basado en descripciones.  
Los pasos restantes serán similares a los anteriores: calcularemos las **similitudes del coseno** y devolveremos las películas más parecidas.

Estos son los pasos que sigo para preparar los datos de géneros y créditos:

1. **Eliminar espacios y convertir todo a minúsculas** en todas las características.  
   De esta manera, nuestro motor no confundirá nombres como **Johnny Depp** y **Johnny Galecki**.  

2. **Mencionar al director tres veces** para darle un mayor peso en comparación con el resto del reparto.


In [38]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [39]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

#### Palabras Clave (Keywords)

Realizaremos una pequeña **preparación previa** de nuestras palabras clave antes de utilizarlas.  
Como primer paso, calcularemos la **frecuencia de aparición** de cada palabra clave que aparece en el conjunto de datos.


In [40]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [41]:
s = s.value_counts()
s[:5]

keyword
independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: count, dtype: int64

Las palabras clave aparecen con frecuencias que van de **1 a 610**.  
No nos sirven aquellas palabras clave que aparecen **solo una vez**, por lo que podemos eliminarlas sin problema.  
Finalmente, convertiremos cada palabra a su **raíz léxica (stem)**, de modo que términos como *Dogs* y *Dog* se consideren equivalentes.


In [42]:
s = s[s > 1]

In [43]:
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [44]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [45]:
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [46]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

 Contamos las veces aparece cada palabra con CountVectorizer sin ponderar las frecuencias:

 Mientras TF-IDF Vectorizer: Pondera frecuencias: Considera qué tan única es la palabra en el anterior.

In [47]:
count = CountVectorizer(analyzer='word', ngram_range=(1, 2), min_df=1, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [48]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [49]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

Reutilizaremos la función **get_recommendations** que escribimos anteriormente.  
Dado que nuestras puntuaciones de **similitud del coseno** han cambiado, esperamos obtener resultados diferentes (y probablemente mejores).  
Verifiquemos nuevamente el caso de **The Dark Knight** y veamos qué recomendaciones obtenemos esta vez.


In [50]:
get_recommendations('The Dark Knight').head(10)

7991         The Dark Knight Rises
6186                 Batman Begins
6587                  The Prestige
2077                     Following
7608                     Inception
4125                      Insomnia
3373                       Memento
8573                  Interstellar
7619    Batman: Under the Red Hood
1122                Batman Returns
Name: title, dtype: object

Podemos estar mucho más satisfechos con los resultados obtenidos esta vez.  
Las recomendaciones parecen haber identificado correctamente otras películas de **Christopher Nolan** (debido al alto peso asignado al director) y las han colocado entre las principales sugerencias.  
Disfruté ver **The Dark Knight**, así como otras películas de la lista, incluyendo **Batman Begins**, **The Prestige** y **The Dark Knight Rises**.

Por supuesto, podemos seguir experimentando con este motor probando diferentes **pesos para nuestras características** (directores, actores, géneros),  
limitando la cantidad de **palabras clave** utilizadas en la mezcla (*soup*),  
ponderando los **géneros según su frecuencia**,  
mostrando solo películas del **mismo idioma**, entre otras posibles mejoras.


Veamos también las recomendaciones para otra película: **Mean Girls**,  
que resulta ser la película favorita de mi novia.



In [51]:
get_recommendations('Mean Girls').head(10)

3311               Head Over Heels
4735                 Freaky Friday
1321              The House of Yes
6245              Just Like Heaven
7865         Mr. Popper's Penguins
7292    Ghosts of Girlfriends Past
6923     The Spiderwick Chronicles
8855                      The DUFF
6662         It's a Boy Girl Thing
7333       I Love You, Beth Cooper
Name: title, dtype: object

#### Popularidad y Calificaciones

Una cosa que notamos acerca de nuestro sistema de recomendación es que sugiere películas **sin tener en cuenta las calificaciones ni la popularidad**.  
Es cierto que **Batman and Robin** comparte muchos personajes con **The Dark Knight**, pero fue una película terrible que no debería recomendarse a nadie.

Por lo tanto, añadiremos un mecanismo para **filtrar las malas películas** y devolver únicamente aquellas que sean **populares** y que hayan tenido una **buena recepción crítica**.

Tomaremos las **25 películas principales** basadas en los puntajes de similitud y calcularemos el número de votos correspondiente al **percentil 60**.  
Luego, usando este valor como \( m \), calcularemos la **calificación ponderada** de cada película utilizando la **fórmula de IMDB**, tal como hicimos en la sección del *Recomendador Simple*.


In [52]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [53]:
improved_recommendations('The Dark Knight')

title  vote_count  vote_average  year  \
7608                           Inception       14075             8  2010   
8573                        Interstellar       11187             8  2014   
6587                        The Prestige        4510             8  2006   
3373                             Memento        4168             8  2000   
7991               The Dark Knight Rises        9263             7  2012   
6186                       Batman Begins        7511             7  2005   
1122                      Batman Returns        1706             6  1992   
132                       Batman Forever        1529             5  1995   
9004  Batman v Superman: Dawn of Justice        7189             5  2016   
1252                      Batman & Robin        1447             4  1997   

            wr  
7608  7.917588  
8573  7.897107  
6587  7.758148  
3373  7.740175  
7991  6.921448  
6186  6.904127  
1122  5.846862  
132   5.054144  
9004  5.013943  
1252  4.287233

Veamos también las recomendaciones para **Mean Girls**.


In [54]:
improved_recommendations('Mean Girls')

title  vote_count  vote_average  year  \
1539                       The Breakfast Club        2189             7  1985   
390                        Dazed and Confused         588             7  1993   
8855                                 The DUFF        1372             6  2015   
3696                     The Princess Diaries        1063             6  2001   
4735                            Freaky Friday         919             6  2003   
6245                         Just Like Heaven         595             6  2005   
6923                The Spiderwick Chronicles         593             6  2008   
7450  American Pie Presents: The Book of Love         454             5  2009   
7292               Ghosts of Girlfriends Past         716             5  2009   
7865                    Mr. Popper's Penguins         775             5  2011   

            wr  
1539  6.709602  
390   6.254682  
8855  5.818541  
3696  5.781086  
4735  5.757786  
6245  5.681521  
6923  5.680901  
7450  5.119690  
7292  5.092422  
7865  5.087912

Desafortunadamente, **Batman and Robin** no desaparece de nuestra lista de recomendaciones.  
Esto probablemente se deba a que tiene una calificación de **4**, que está solo ligeramente por debajo del promedio en TMDB.  
Ciertamente no merece un 4 cuando películas increíbles como **The Dark Knight Rises** apenas alcanzan un **7** (cuestión de gustos claro,jejeje).  
Sin embargo, no hay mucho que podamos hacer al respecto.

Por lo tanto, concluiremos aquí nuestro **Recomendador Basado en Contenido**.


## Filtrado Colaborativo

El recomendador basado en contenido muestra algunas limitaciones importantes. Solo es capaz de sugerir películas que son *similares* a una película en concreto. Es decir, no es capaz de capturar gustos y proporcionar recomendaciones a través de diferentes géneros.

Además, el recomendador que construimos no es realmente personal, ya que no captura los gustos personales y los sesgos de un usuario. Cualquier persona que lo consulte para obtener recomendaciones basadas en una película recibirá las mismas recomendaciones para esa película, sin importar quién sea.

Por lo tanto, en esta sección, usaremos una técnica llamada **Filtrado Colaborativo** para hacer recomendaciones a los espectadores. El Filtrado Colaborativo se basa en la idea de que usuarios similares a uno pueden usarse para predecir cuánto me gustará un producto o servicio particular que esos usuarios han usado/experimentado pero yo no.

No vamos a implementar el Filtrado Colaborativo desde cero. En su lugar, usaremos la librería **Surprise**, que utiliza algoritmos extremadamente potentes como la **Descomposición en Valores Singulares (SVD)** para minimizar el RMSE (Error Cuadrático Medio) y dar mejores recomendaciones.

## Descomposición en Valores Singulares (SVD)

La **Descomposición en Valores Singulares (SVD)** es una técnica fundamental del álgebra lineal que descompone una matriz en tres matrices componentes. Para el contexto de sistemas de recomendación, se aplica a la matriz usuario-ítem para descubrir patrones latentes en los datos.

### Representación Matemática

Dada una matriz A de dimensiones m×n, la SVD la factoriza como:

**A = U × Σ × Vᵀ**

Donde:
- **U**: Matriz ortogonal m×m (vectores singulares izquierdos)
- **Σ**: Matriz diagonal m×n (valores singulares en orden descendente)
- **Vᵀ**: Matriz ortogonal n×n transpuesta (vectores singulares derechos)

### Aplicación en Sistemas de Recomendación

En filtrado colaborativo, la SVD se utiliza para:

- **Reducir dimensionalidad**: Mantener solo los k valores singulares más importantes
- **Capturar factores latentes**: Los valores singulares representan "factores" o "características ocultas" que explican las preferencias de los usuarios
- **Predecir calificaciones**: Completar los valores faltantes en la matriz usuario-ítem

### Ventajas en Surprise

La librería Surprise optimiza SVD para:
- Manejar datos dispersos eficientemente
- Minimizar el error de predicción (RMSE)
- Escalar a grandes conjuntos de datos
- Proporcionar recomendaciones personalizadas y precisas

In [55]:
#Tomados el rating de 1 a 5 estrellas
reader = Reader(rating_scale=(1, 5))

In [56]:
#
ratings = pd.read_csv('data/ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [57]:
#Cargamos las calificaciones de los usuarios a las pelis
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

In [58]:
np.random.seed(42)

In [59]:
svd = SVD(random_state=42)
results = cross_validate(svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True) # cv=5, usa 5-fold cross validation (divide datos en 5 partes, usa 4 para entrenar y 1 para test)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8972  0.8918  0.8992  0.8940  0.9023  0.8969  0.0037  
MAE (testset)     0.6909  0.6880  0.6933  0.6871  0.6953  0.6909  0.0031  
Fit time          0.39    0.46    0.38    0.37    0.38    0.39    0.03    
Test time         0.04    0.13    0.04    0.04    0.10    0.07    0.04    


Obtenemos un **Error Cuadrático Medio** promedio de 0.8970, lo cual es más que suficiente para nuestro caso. Procedamos ahora a entrenar con nuestro dataset completo y generar las predicciones.

In [60]:
trainset = data.build_full_trainset() #Se usa ntodos los datos
svd.fit(trainset)

*Actividad 1*

Selecciona al usuario 1 y revisa las calificaciones que ha dado.

In [61]:
# Completa. Usa ratings[]
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [62]:
# Genera una Predicción para el usuario 1 y ve si puso 3 de calificación en la película 302
# usa svd.predict ()

prediction = svd.predict(uid=1, iid=302, r_ui=3)
print(prediction)

user: 1          item: 302        r_ui = 3.00   est = 2.72   {'was_impossible': False}


Por ejemplo puedes obtener esto para la película 302

Prediction(uid=1, iid=302, r_ui=3, est=2.5027733276596384, details={'was_impossible': False})

Para la película con ID 302, obtenemos una predicción estimada de **2.502**. Una característica sorprendente de este sistema de recomendación es que no le importa qué sea la película (o qué contenido tenga). Funciona puramente en base a un ID de película asignado e intenta predecir las calificaciones basándose en cómo los otros usuarios han calificado la película.

## Recomendador híbrido

![](https://www.toonpool.com/user/250/files/hybrid_20095.jpg)

En esta sección, se va a construir un recomendador híbrido simple que reúna las técnicas que hemos implementado en los motores basados en contenido y en filtrado colaborativo. Así es como funcionará:

* **Entrada:** ID de Usuario y el Título de una Película
* **Salida:** Películas similares ordenadas en base a las calificaciones esperadas por ese usuario en particular.

In [63]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [64]:
id_map = pd.read_csv('data/links_small.csv')[['movieId', 'tmdbId']]
id_map['tmdbId'] = id_map['tmdbId'].apply(convert_int)
id_map.columns = ['movieId', 'id']
id_map = id_map.merge(smd[['title', 'id']], on='id').set_index('title')

In [65]:
indices_map = id_map.set_index('id')

In [66]:
def hybrid(userId, title):
    idx = indices[title]
    tmdbId = id_map.loc[title]['id']
    #print(idx)
    movie_id = id_map.loc[title]['movieId']
    
    sim_scores = list(enumerate(cosine_sim[int(idx)]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year', 'id']]
    movies['est'] = movies['id'].apply(lambda x: svd.predict(userId, indices_map.loc[x]['movieId']).est)
    movies = movies.sort_values('est', ascending=False)
    return movies.head(10)

Prueba Avatar en dos usuarios, el 1 y alguno más

In [67]:
hybrid(1, 'Avatar')

title  vote_count  vote_average  year  \
8357             Star Trek Into Darkness      4479.0           7.4  2013   
962                               Aliens      3282.0           7.7  1986   
999                       The Terminator      4208.0           7.4  1984   
8622          X-Men: Days of Future Past      6155.0           7.5  2014   
522           Terminator 2: Judgment Day      4274.0           7.7  1991   
2006                    Fantastic Planet       140.0           7.6  1973   
1613  Darby O'Gill and the Little People        35.0           6.7  1959   
3052     Sinbad and the Eye of the Tiger        39.0           6.3  1977   
4323      Piranha Part Two: The Spawning        41.0           3.9  1981   
1660          Return from Witch Mountain        38.0           5.6  1978   

          id       est  
8357   54138  3.064241  
962      679  3.015366  
999      218  3.006598  
8622  127585  2.980404  
522      280  2.875950  
2006   16306  2.820924  
1613   18887  2.728216  
3052   11940  2.712663  
4323   31646  2.692581  
1660   14822  2.676243

In [68]:
hybrid(2, 'Avatar')

title  vote_count  vote_average  year      id  \
522        Terminator 2: Judgment Day      4274.0           7.7  1991     280   
8357          Star Trek Into Darkness      4479.0           7.4  2013   54138   
962                            Aliens      3282.0           7.7  1986     679   
8622       X-Men: Days of Future Past      6155.0           7.5  2014  127585   
344                         True Lies      1138.0           6.8  1994   36955   
2006                 Fantastic Planet       140.0           7.6  1973   16306   
999                    The Terminator      4208.0           7.4  1984     218   
1660       Return from Witch Mountain        38.0           5.6  1978   14822   
3052  Sinbad and the Eye of the Tiger        39.0           6.3  1977   11940   
4001                  Hawk the Slayer        13.0           4.5  1980   25628   

           est  
522   4.112434  
8357  4.027732  
962   3.809366  
8622  3.739850  
344   3.736516  
2006  3.669347  
999   3.649392  
1660  3.598725  
3052  3.563369  
4001  3.547432

In [69]:
hybrid(3, 'Avatar')

title  vote_count  vote_average  year  \
999                       The Terminator      4208.0           7.4  1984   
962                               Aliens      3282.0           7.7  1986   
522           Terminator 2: Judgment Day      4274.0           7.7  1991   
1613  Darby O'Gill and the Little People        35.0           6.7  1959   
1368                             Titanic      7770.0           7.5  1997   
8357             Star Trek Into Darkness      4479.0           7.4  2013   
2006                    Fantastic Planet       140.0           7.6  1973   
2124                         Superman II       642.0           6.5  1980   
910                            The Abyss       822.0           7.1  1989   
4001                     Hawk the Slayer        13.0           4.5  1980   

         id       est  
999     218  3.770350  
962     679  3.753860  
522     280  3.670539  
1613  18887  3.570995  
1368    597  3.561213  
8357  54138  3.553391  
2006  16306  3.512679  
2124   8536  3.467287  
910    2756  3.418567  
4001  25628  3.386750

Podemos observar que para nuestro recomendador híbrido, obtenemos recomendaciones diferentes para distintos usuarios aunque la película de entrada sea la misma. Por lo tanto, nuestras recomendaciones son más personalizadas y adaptadas a usuarios particulares.

## Conclusión

En este notebook, hemos construido 4 sistemas de recomendación diferentes basados en distintas ideas y algoritmos. Estos son los siguientes:

1. **Recomendador Simple:** Este sistema utilizó el Conteo de Votos y los Promedios de Calificación generales de TMDB para crear Listas de las Mejores Películas, tanto en general como para un género específico. Se utilizó el Sistema de Ponderación de Calificaciones de IMDB para calcular las puntuaciones sobre las cuales finalmente se realizó la ordenación.

2. **Recomendador Basado en Contenido:** Construimos dos motores basados en contenido; uno que tomó como entrada la sinopsis y las taglines de las películas, y otro que utilizó metadatos como elenco, equipo, género y palabras clave para generar predicciones. También implementamos un filtro simple para dar mayor preferencia a las películas con más votos y mejores calificaciones.

3. **Filtrado Colaborativo:** Utilizamos la potente librería Surprise para construir un filtro colaborativo basado en descomposición en valores singulares. El RMSE obtenido fue menor que 1 y el motor proporcionó calificaciones estimadas para un usuario y película dados.

4. **Motor Híbrido:** Combinamos ideas del filtrado basado en contenido y colaborativo para construir un motor que proporcionó sugerencias de películas a un usuario particular basándose en las calificaciones estimadas que había calculado internamente para ese usuario.